### ISSUE TITLE
Re-raising Failure in 8.0.1 causes exception in pb.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3174]
|<img alt="aquamatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| aquamatt reported|
|-|-|
|Trac ID|trac#3174|
|Type|release blocker: regression|
|Created|2008-04-10 09:06:31Z|
|Branch|https://github.com/twisted/twisted/tree/copied-copied-copyable-failure-3174|

Files attached: 
    * my patch to pb.py to resolve this issue
    * script to demonstrate the issue and test the fix

When making spread.pb calls via a chain of processes, Failure instances may not traverse all nodes.

Case example:
Process 1 raises an exception which gets wrapped in a Failure and passed to calling process
Process 2 Failure appears in errback handler; chooses to put it on to the errback call for a upstream process
Process 3 SHOULD receive the Failure

In 8.0.1 exception raised in Process 2 as follows:
```
Unhandled error in Deferred:
Traceback (most recent call last):
  File "/home/mp/python/t8/lib/python/twisted/internet/defer.py", line 323, in _runCallbacks
    self.result = callback(self.result, *args, **kw)
  File "testerr.py", line 64, in callerror
    rd.errback(err)
  File "/home/mp/python/t8/lib/python/twisted/internet/defer.py", line 264, in errback
    self._startRunCallbacks(fail)
  File "/home/mp/python/t8/lib/python/twisted/internet/defer.py", line 307, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/home/mp/python/t8/lib/python/twisted/internet/defer.py", line 323, in _runCallbacks
    self.result = callback(self.result, *args, **kw)
  File "/home/mp/python/t8/lib/python/twisted/spread/pb.py", line 904, in _sendFailureOrError
    self._sendFailure(fail, requestID)
  File "/home/mp/python/t8/lib/python/twisted/spread/pb.py", line 913, in _sendFailure
    self._sendError(fail, requestID)
  File "/home/mp/python/t8/lib/python/twisted/spread/pb.py", line 928, in _sendError
    self.sendCall("error", requestID, self.serialize(fail))
  File "/home/mp/python/t8/lib/python/twisted/spread/pb.py", line 772, in serialize
    return jelly(object, self.security, None, self)
  File "/home/mp/python/t8/lib/python/twisted/spread/jelly.py", line 1122, in jelly
    return _Jellier(taster, persistentStore, invoker).jelly(object)
  File "/home/mp/python/t8/lib/python/twisted/spread/jelly.py", line 476, in jelly
    return obj.jellyFor(self)
  File "/home/mp/python/t8/lib/python/twisted/spread/flavors.py", line 304, in jellyFor
    state = self.getStateToCopyFor(p)
  File "/home/mp/python/t8/lib/python/twisted/spread/flavors.py", line 272, in getStateToCopyFor
    return self.getStateToCopy()
  File "/home/mp/python/t8/lib/python/twisted/spread/pb.py", line 425, in getStateToCopy
    state['type'] = reflect.qual(self.type) # Exception class
  File "/home/mp/python/t8/lib/python/twisted/python/reflect.py", line 296, in qual
    return clazz.__module__ + '.' + clazz.__name__
exceptions.AttributeError: 'str' object has no attribute '__module__'
```

Attachments:

* [pb_8_0_1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a2b/a2bd59bb95d829ebae61ac6732f7f2dfa8198262/2b5522d1d8f7a6a8fd0ecb41b34ee6b43719ad5f.patch) (636 bytes) - added by aquamatt on 2008-04-10 09:06:57Z - Patch to pb.py for this issue
* [test_patch_8_0_1.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a2b/a2bd59bb95d829ebae61ac6732f7f2dfa8198262/c8fafe827fd8228d09ee8376472ed1b484ed5500.py) (3473 bytes) - added by aquamatt on 2008-04-10 09:08:22Z - Test script to demonstrate issue

<details><summary>Searchable metadata</summary>

```
trac-id__3174 3174
type__release_blocker__regression release blocker: regression
reporter__aquamatt aquamatt
priority__highest highest
milestone__ 
branch__branches_copied_copied_copyable_failure_3174 branches/copied-copied-copyable-failure-3174
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1207818391000000 1207818391000000
changetime__1207859794000000 1207859794000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

Thanks for the report. FWIW, the problem (if this is one) was inserted in r22574.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

I have a solution that looks much cleaner, but I don't know if it's practical:

```
Index: twisted/spread/pb.py
===================================================================
--- twisted/spread/pb.py        (revision 23266)
+++ twisted/spread/pb.py        (working copy)
@@ -441,6 +441,16 @@
     printBriefTraceback = printTraceback
     printDetailedTraceback = printTraceback

+
+    def setCopyableState(self, state):
+        klass = reflect.namedAny(state['type'])
+        inst = klass(state['value'])
+        state['value'] = inst
+        state['type'] = klass
+        self.__dict__ = state
+
+
+
 setUnjellyableForClass(CopyableFailure, CopiedFailure)

 def failure2Copyable(fail, unsafeTracebacks=0):
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

New in 8.0.1

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I'm not sure about that `reflect.namedAny`.  At least something should probably be checking `isModuleAllowed`.  I'm inclined to return to the old behavior in the simplest way, even though it involves adding an `isinstance` call which is lame.  (I don't much like `CopyableFailure`.)

**Comment by twisted-trac** [comment]
|<img alt="aquamatt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|aquamatt commented|
|-|-|

therve's patch is neater than mine and works for me, for what it's worth! I'm running with it for the moment but obviously am unsure what side-effects may occur.

Thanks for the quick discussion :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [23293]) Branching to 'copied-copied-copyable-failure-3174'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

I did the type-check based fix and added a couple unit tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Changed the unit tests slightly to explicitly construct a new `CopyableFailure` rather than using `raiseException`.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @exarkun|
|-|-|

Please wrap the tests to 80 columns, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [23300]) Merge copied-copied-copyable-failure-3174

Author: exarkun, aquamatt
Reviewer: therve
Fixes [#3174](https://github.com/twisted/twisted/issues/3174)

Add a test for jellying a CopyableFailure created from a CopiedFailure
and change the CopyableFailure jellying logic to (once again) support
this case by allowing string exception types to pass through unmodified.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
